tests/test_TFUtil.py

```diff
@@ -469,6 +469,15 @@ def test_Data_copy_compatible_to_batch_feature_is_dynamic():
   assert_equal(t.get_size_dim_tag(0), dec.get_time_dim_tag())
 
 
+def test_Data_get_common_data_different_time_dim():
+  d1 = Data(name='t', shape=(None, 32, 128), dtype='float32', auto_create_placeholders=True)
+  d2 = Data(name='r', shape=(None, 32, 128), dtype='float32', auto_create_placeholders=True)
+  common_data = Data.get_common_data([d1, d2], warnings_out=sys.stdout)
+  op1 = d1.copy_compatible_to(common_data).placeholder
+  op2 = d2.copy_compatible_to(common_data).placeholder
+  assert op1.shape == op2.shape
```

Comment:
The `assert` is wrong. In one case, you get sth like (None, 1, 32, 128), in the other case, you get sth like (1, None, 32, 128), so they will not be the same (the purpose of `copy_compatible_to` is to make them compatible, not to make them the same).
Btw, is it on purpose that you take the `placeholder` here? Why not just the `Data` instance directly? Please do both in the code, and add a comment why you do the `placeholder` check.

Reply:
See comment above, this is the behavior of the `CombineLayer`. In particular, this method:
https://github.com/rwth-i6/returnn/blob/68ee06301cd7597ba3909d8a5baebfbc27a72308/TFNetworkLayer.py#L5598
`  def _op_dense_fn(sources, fn):
    """
    :param list[LayerBase] sources:
    :param ((x1,x2) -> y) fn: function to perform on x1 and x2
    :rtype: tf.Tensor
    """
    # Earlier we checked for the same dense dim.
    # Now, we completely rely on Data.get_common_data and Data.copy_compatible_to.
    # That should fail if they are not compatible. Otherwise it would add any needed broadcast dimensions.
    # All the dense element-wise functions should be able to deal with broadcasting.
    common_data = Data.get_common_data([s.output for s in sources])
    x = sources[0].output.copy_compatible_to(common_data).placeholder
    for source in sources[1:]:
      x2 = source.output.copy_compatible_to(common_data).placeholder
      x = fn(x, x2)
    return x`

Reply:
The assert is wrong. In one case, you get sth like (None, 1, 32, 128), in the other case, you get sth like (1, None, 32, 128), so they will not be the same (the purpose of copy_compatible_to is to make them compatible, not to make them the same).
Btw, is it on purpose that you take the placeholder here? Why not just the Data instance directly? Please do both in the code, and add a comment why you do the placeholder check.

Reply:
Is this test case still needed?

Reply:
I think no, I'll delete it